File: benchmark/src/nodejs/union_find.js
const n=200000,U=400000,parent=[...Array(n).keys()],size=Array(n).fill(1)
function find(x){while(parent[x]!=x)x=parent[x];return x}
function union(a,b){let ra=find(a),rb=find(b);if(ra!=rb){if(size[ra]<size[rb]){parent[ra]=rb;size[rb]+=size[ra]}else{parent[rb]=ra;size[ra]+=size[rb]}}}
let seed=12345;function ri(m){seed=(seed*16807)%2147483647;return seed%m}
const start=Date.now()
for(let i=0;i<U;i++)union(ri(n),ri(n))
let roots=0;for(let i=0;i<n;i++)if(parent[i]===i)roots++
const end=Date.now();console.log("Result: "+roots);console.log("Time: "+Math.floor(end-start)+"ms")

File: benchmarks/src/perl/union_find.pl
use Time::HiRes qw(time);
my $n = 200000; my $U = 400000;
my @parent = (0..$n-1); my @size = (1) x $n;
sub find { my $x = shift; while ($parent[$x] != $x) { $x = $parent[$x] } $x }
sub union {
    my ($a, $b) = @_; my $ra = find($a); my $rb = find($b); return if $ra == $rb;
    if ($size[$ra] < $size[$rb]) { $parent[$ra] = $rb; $size[$rb] += $size[$ra] }
    else { $parent[$rb] = $ra; $size[$ra] += $size[$rb] }
}
my $seed = 12345;
sub ri { my $m = shift; $seed = ($seed * 16807) % 2147483647; $seed % $m }
my $start = time();
for (1..$U) { union(ri($n), ri($n)) }
my $roots = 0;
for (0..$n-1) { $roots++ if $parent[$_] == $_ }
my $end = time(); my $ms = ($end - $start) * 1000;
print "Result: $roots\n"; printf "Time: %.0fms\n", $ms;
